packages/auth/lib/Users.js

```diff
@@ -40,9 +41,84 @@ const {
   FRONTEND_BASE_URL
 } = process.env
 
-const signIn = async (_email, context, pgdb, req, consents) => {
+// in order of preference
+const enabledFirstFactors = async (email, pgdb) => {
+  const userWithDevices = await pgdb.query(`
+    SELECT
+      u.*,
+      jsonb_agg(d.*) as devices
+    FROM
+      users u
+    LEFT JOIN
+      devices d
+      ON u.id = d."userId"
+    WHERE
+      u.email = :email
+    GROUP BY
+      u.id
+  `, {
+    email
+  })
+    .then(result => {
+      const user = result[0]
+      if (!user) {
+        return user
+      }
+      return {
+        ...user,
+        devices: user.devices
+          .filter(Boolean)
+      }
+    })
+
+  const { APP, EMAIL_TOKEN } = TokenTypes
+
+  const enabledTokenTypes = [EMAIL_TOKEN]
+  if (userWithDevices && userWithDevices.devices.length > 0) {
+    enabledTokenTypes.push(APP)
+  }
+
+  const { preferredFirstFactor } = userWithDevices || {}
+  return enabledTokenTypes
+    .sort((a, b) => {
```

Comment:
Suggestion: Since `EMAIL_TOKEN` is declared to be last, I'd simplify this sort thing as follows:

```js
const enabledTokenTypes = []

// [...]

// Push factors back if not preferred
enabledTokenTypes.sort((a, b) => b === preferredFirstFactor ? 1 : 0)

// Append EMAIL_TOKEN
if (enabledTokenTypes.includes(EMAIL_TOKEN)) {
  enabledTokenTypes.push(EMAIL_TOKEN)
}

return enabledTokenTypes
```

What throw me off is `// EMAIL_TOKEN always last`, it seems to contradict `// preferred always wins`.

An email token can only "always be last"  if `EMAIL_TOKEN` won't be a `preferredFirstFactor`.

From what I skimmed in the code, it might as well be possible that `EMAIL_TOKEN` is a preferred first factor. Is that accurate?